proof-systems/src/darlin/poly-commit/src/constraints/mod.rs

```diff
@@ -0,0 +1,447 @@
+        proof: &Self::MultiPointProof,
+        random_oracle: &mut Self::RandomOracle,
+    ) -> Result<Self::VerifierState, Self::Error> {
+        let commitment_map: BTreeMap<_, _> = labeled_commitments
+            .into_iter()
+            .map(|commitment| (commitment.label(), commitment.commitment()))
+            .collect();
+
+        let lambda_bits = random_oracle.enforce_get_challenge::<_, 128>(
+            cs.ns(|| "squeezing random challenge for multi-point-multi-poly verify"),
+            )?;
+
+        let lambda = Self::challenge_to_non_native_field_element(
+            cs.ns(|| "convert lambda to non native field gadget"),
+            &lambda_bits,
+        )?;
+
+        random_oracle.enforce_record(
+            cs.ns(|| "absorb commitment to polynomial h"),
+            proof.get_h_commitment().clone(),
+        )?;
+        let evaluation_point_bits = random_oracle.enforce_get_challenge::<_, 128>(
+            cs.ns(|| "squeeze evaluation point for multi-point multi-poly verify"),
+            )?;
+        let evaluation_point = Self::challenge_to_non_native_field_element(
+            cs.ns(|| "evaluation point from squeezed bits"),
+            &evaluation_point_bits,
+        )?;
+
+        let mut points_iter = points.iter().rev();
+        let zero_value = NonNativeFieldGadget::<G::ScalarField, ConstraintF>::zero(&mut cs)?;
+        let default_label = (String::from("default label"), String::from("default label"));
+        let ((label, point_label), point) =
+            points_iter.next().unwrap_or((&default_label, &zero_value));
+        let commitment = *commitment_map
+            .get(label)
+            .ok_or(SynthesisError::Other(String::from(format!(
+                "commitment with label {} not found",
+                label
+            ))))?;
+        let value =
+            values
+                .get(&(label.clone(), point_label.clone()))
+                .ok_or(SynthesisError::Other(String::from(format!(
+                    "evaluation for point {} not found",
+                    point_label
+                ))))?;
+        let z_i_over_z_value = evaluation_point
+            .sub(cs.ns(|| "evaluation_point - point for last point"), &point)?
+            .inverse(cs.ns(|| "(evaluation_point - point)^-1 for last point"))?;
+        let z_i_over_z_bits = z_i_over_z_value
+            .to_bits_for_normal_form(cs.ns(|| "z_i_over_z_value to bits for last point"))?;
+
+        let mut batched_commitment = safe_mul_bits::<ConstraintF, G, PC, Self, _, _>(
+            cs.ns(|| "commitment*z_i_over_z for last point"),
+            &commitment,
+            z_i_over_z_bits.iter().rev(),
+        )?; // reverse order of bits since mul_bits requires little endian representation
+        let mut batched_value = value.mul(
+            cs.ns(|| "value*z_i_over_z for last point"),
+            &z_i_over_z_value,
+        )?;
+
+        for ((label, point_label), point) in points_iter {
```

Comment:
It's not clear to me why you bothered performing the first iteration of the loop outside the for. Can't you do in the same way as for the single_point_multi_poly ?

Reply:
Actually when merging the changes from `sponge` I have to change the structure of the function to deal with the new implementation of `QueryMap`. I also remove code duplication with this update.